File: 2021/go/day01.go
package aoc2021

import (
	"fmt"

	"github.com/elliotchance/pie/pie"
	"github.com/scottames/adventofcode/pkg/helpers"
)

const (
	year = 2021
)

// Day01 - Day 1 Part 1 & 2
func Day01() error {
	get, err := helpers.ReadInput(year, 1)
	if err != nil {
		return err
	}

	input, err := helpers.StringSplitNewlinesToInts(get)
	if err != nil {
		return err
	}

	helpers.PrintPart1()
	fmt.Println("Number of increases: ", increasesInList(input))

	helpers.PrintPart2()
	fmt.Println("Number of three sequential increases: ", TripleSequenceIncreases(input))

	return nil
}

func increasesInList(ints pie.Ints) int {
	p := ints.First()
	count := 0

	ints.Filter(
		func(i int) bool {
			if i > p {
				count++
			}
			p = i

			return false
		},
	)

	return count
}

func TripleSequenceIncreases(ints pie.Ints) int {
	p := 0
	count := 0
	for i := 0; i < len(ints)-2; i++ {
		s := ints[i : i+3].Sum()
		if i > 0 && s > p {
			count++
		}
		p = s
	}

	return count
}

File: 2021/python/day01.py
#!/bin/env python3

from typing import List
from utils.print import print_part

DAY = "day1"
INPUT = "input/2021/" + DAY + ".txt"


def main():
    list = parse()

    print_part(1)
    print("Number of increases: ", str(increases_in_list(list)))
    print()

    print_part(2)
    print(
        "Number of three sequential increases: ",
        str(triple_sequential_increases_in_list(list)),
    )
    print()


def parse() -> List[int]:
    ints = []
    with open(INPUT, "r") as txt:
        ints = [int(i) for i in txt.read().rstrip().split("\n")]

    return ints


def increases_in_list(ints: List[int]) -> int:
    result = [n for i, n in enumerate(ints) if i != 0 and n > ints[i - 1]]

    return len(result)


def triple_sequential_increases_in_list(ints: List[int]) -> int:
    return increases_in_list([a + b + c for a, b, c in zip(ints, ints[1:], ints[2:])])


if __name__ == "__main__":
    main()
